diskqueue.go

```diff
@@ -395,6 +422,188 @@ func (d *diskQueue) readOne() ([]byte, error) {
 	return readBuf, nil
 }
 
+func (d *diskQueue) removeReadFile() error {
+	var err error
+
+	if d.readFileNum == d.writeFileNum {
+		d.skipToNextRWFile()
+		return nil
+	}
+
+	if d.readFile == nil {
+		curFileName := d.fileName(d.readFileNum)
+
+		d.readFile, err = os.OpenFile(curFileName, os.O_RDONLY, 0600)
+		if err != nil {
+			return err
+		}
+
+		d.reader = bufio.NewReader(d.readFile)
+	}
+
+	closeReadFile := func() {
+		d.readFile.Close()
+		d.readFile = nil
+	}
+	defer closeReadFile()
+
+	// read total messages number at the end of the file
+	_, err = d.readFile.Seek(-numFileMsgBytes, 2)
+	if err != nil {
+		return err
+	}
+
+	var totalMessages int64
+	err = binary.Read(d.reader, binary.BigEndian, &totalMessages)
+	if err != nil {
+		return err
+	}
+
+	// update depth with the remaining number of messages
+	d.depth -= totalMessages - d.readMessages
+
+	// we have not finished reading this file
+	if d.readFileNum == d.nextReadFileNum {
+		d.nextReadFileNum++
+		d.nextReadPos = 0
+	}
+
+	d.moveToNextReadFile()
+
+	return nil
+}
+
+// walk through all of the files in the DiskQueue directory
+func (d *diskQueue) walkDiskQueueDir(fn func(os.FileInfo) error) error {
+	fileInfos, err := ioutil.ReadDir(d.dataPath)
+
+	if err != nil {
+		return err
+	}
+
+	for _, fileInfo := range fileInfos {
+		// only go through files and skip directories
+		if !fileInfo.IsDir() {
+			err = fn(fileInfo)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (d *diskQueue) getAllBadFileInfo() ([]os.FileInfo, error) {
+	var badFileInfos []os.FileInfo
+
+	getAllBadFileInfo := func(fileInfo os.FileInfo) error {
+		// only accept "bad" files created by this DiskQueue object
+		if badFileNameRegexp.MatchString(fileInfo.Name()) {
+			badFileInfos = append(badFileInfos, fileInfo)
+		}
+
+		return nil
+	}
+
+	err := d.walkDiskQueueDir(getAllBadFileInfo)
+
+	return badFileInfos, err
+}
+
+// get the accurate total non-"bad" file size
+func (d *diskQueue) updateTotalDiskSpaceUsed() error {
+	d.totalDiskSpaceUsed = maxMetaDataFileSize
+
+	updateTotalDiskSpaceUsed := func(fileInfo os.FileInfo) error {
+		// only accept files created by this DiskQueue object
+		if fileNameRegexp.MatchString(fileInfo.Name()) || badFileNameRegexp.MatchString(fileInfo.Name()) {
+			d.totalDiskSpaceUsed += fileInfo.Size()
+		}
+
+		return nil
+	}
+
+	return d.walkDiskQueueDir(updateTotalDiskSpaceUsed)
+}
+
+func (d *diskQueue) freeDiskSpace(expectedBytesIncrease int64) error {
+	var err error
+	var badFileInfos []os.FileInfo
+
+	badFileInfos, err = d.getAllBadFileInfo()
+	if err != nil {
+		d.logf(ERROR, "DISKQUEUE(%s) failed to retrieve all .bad file info - %s", d.name, err)
+	}
+
+	// keep freeing up disk space until we have enough space to write this message
+	for d.totalDiskSpaceUsed+expectedBytesIncrease > d.maxBytesDiskSpace {
```

Comment:
Check whether all files are deleted.

Reply:
Wouldn't that be the same as checking if the totalDiskSpaceUsed+increase > 0?

Reply:
Is this to prevent the case where all files are deleted, but there is no space to write the message added or in other words when expectedBytesIncrease > d.maxBytesDiskSpace - maxMetaDataFileSize (causing an infinite loop)? If this is the case, should I check whether expectedBytesIncrease > d.maxBytesDiskSpace - maxMetaDataFileSize in `checkDiskSpace`?

Reply:
Updated in https://github.com/kev1n80/go-diskqueue/pull/6/files#r665626211

Reply:
Good catch, I updated the code.